0 또는 양의 정수가 주어졌을 때, 정수를 이어 붙여 만들 수 있는 가장 큰 수를 알아내 주세요.

예를 들어, 주어진 정수가 [6, 10, 2]라면 [6102, 6210, 1062, 1026, 2610, 2106]를 만들 수 있고, 이중 가장 큰 수는 6210입니다.

0 또는 양의 정수가 담긴 배열 numbers가 매개변수로 주어질 때, 순서를 재배치하여 만들 수 있는 가장 큰 수를 문자열로 바꾸어 return 하도록 solution 함수를 작성해주세요.

제한 사항
- numbers의 길이는 1 이상 100,000 이하입니다.
- numbers의 원소는 0 이상 1,000 이하입니다.
- 정답이 너무 클 수 있으니 문자열로 바꾸어 return 합니다.

In [1]:
def solution(numbers):
    answer = ''
    numbers = [str(n) for n in numbers]
    numdict = dict()
    for num in numbers:
        if num[0] not in numdict:
            numdict.update({num[0]:[num]})
        else:
            numdict.get(num[0]).append(num)

    for key in numdict:
        if len(numdict.get(key)) > 1:
            tmp = numdict.get(key)
            numdict.update({key:[]})
            while tmp:
                m = '0'
                for n in tmp:
                    if n[0] == m[0]:
                        m = n if n + m >= m + n else m
                    elif n[0] > m[0]:
                        m = n
                n = tmp.pop(tmp.index(m))
                numdict.get(key).append(n)
    
    for key in sorted(numdict.keys(),reverse=True):
        answer += ''.join(numdict.get(key))
    
    return answer if answer != '0'*len(numbers) else '0'

In [2]:
solution([6, 10, 2])

'6210'

In [3]:
solution([3,30,34,5,9])

'9534330'

In [4]:
solution([9,382,38,1])

'9383821'

In [5]:
solution([12,121])

'12121'

In [6]:
solution([0,0,0,0])

'0'

In [7]:
solution([0,0,0,1000])

'1000000'

In [8]:
solution([2,22])

'222'

In [9]:
solution([41,415])

'41541'

이렇게 풀었는데 5개의 시간초과가 있었다.

두번째 for문의 시간복잡도가 O(n^2)이어서 그부분을 고쳐본다.

In [10]:
def find_max(a,b):
    return a <= b if len(a) == len(b) else a+b <= b+a

def quick_sort(arr):
    l = len(arr)
    if l <= 1:
        return arr
    else:
        i = arr[0]
        greater = [e for e in arr[1:] if find_max(e,i)]
        less = [e for e in arr[1:] if not find_max(e,i)]
        return quick_sort(less) + [i] + quick_sort(greater)
    
    
def solution(numbers):
    if set(numbers) == {0}:
        return '0'
    
    numbers = [str(n) for n in numbers]
    numbers = quick_sort(numbers)
    
    return ''.join(numbers)

In [11]:
solution([9,382,38,1])

'9383821'

In [12]:
solution([0,0,0,0])

'0'

In [13]:
solution([0,0,0,1000])

'1000000'

고치려다가 다른 코드를 참고해서 다시 짰다.

정렬에 대한 개념을 다시 잡아야겠다!

다른 코드를 보니 그냥 sort함수를 가지고 짜는데, key를 lambda x : x\*3 이런 식으로 했다.(x는 0이상 1000이하)

예를들어 [2,23]인 경우 222와 232323을 비교하는 것이다. 애초에 앞의 3자리만 비교하면 되므로 이렇게 풀어도 상관 없다.

이렇게 생각하긴 했는데 lambda사용법이 익숙하지 않아서 못풀었다!

In [14]:
def solution(numbers):
    numbers = list(map(str, numbers))
    numbers.sort(key=lambda x: x*3, reverse=True)
    return str(int(''.join(numbers)))

In [15]:
solution([2,23])

'232'